File: Java/src/utils/FileHandler.java
package utils;

import model.Employee;
import view.ContentView;
import java.io.*;
import java.util.ArrayList;

/**
 * FileHandler demonstrates parallel programming because it implements Runnable.
 * A separate thread is used for file reading or writing.
 */
public class FileHandler implements Runnable {
    private String filename;
    private ArrayList<Employee> employees;
    private boolean isReading;
    private ContentView view;

    public FileHandler(String filename, ArrayList<Employee> employees, boolean isReading, ContentView view) {
        this.filename = filename;
        this.employees = employees;
        this.isReading = isReading;
        this.view = view;
    }

    @Override
    public void run() {
        if (isReading) {
            readFromFile();
        } else {
            writeToFile();
        }
    }

    private void readFromFile() {
        view.showMessage("[THREAD-" + Thread.currentThread().getId() + "] Reading from file: " + filename);
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = reader.readLine()) != null) {
                view.showMessage("Read: " + line);
            }
            view.showMessage("[THREAD] File reading completed successfully");
        } catch (FileNotFoundException e) {
            view.showError("File not found: " + e.getMessage());
        } catch (IOException e) {
            view.showError("IO Error while reading: " + e.getMessage());
        }
    }

    private void writeToFile() {
        view.showMessage("[THREAD-" + Thread.currentThread().getId() + "] Writing to file: " + filename);
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (Employee emp : employees) {
                writer.write(emp.toString());
                writer.newLine();
            }
            view.showMessage("[THREAD] File writing completed successfully. " + employees.size() + " records saved.");
        } catch (IOException e) {
            view.showError("IO Error while writing: " + e.getMessage());
        }
    }
}

File: CSharp/Utils/FileHandler.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using EmployeeManagementSystem.Models;
using EmployeeManagementSystem.Views;

namespace EmployeeManagementSystem.Utils
{
    // FileHandler demonstrates multithreading because Run() is executed by a separate Thread.
    public class FileHandler
    {
        private string filename;
        private List<Employee> employees;
        private bool isReading;
        private ContentView view;

        public FileHandler(string filename, List<Employee> employees, bool isReading, ContentView view)
        {
            this.filename = filename;
            this.employees = employees;
            this.isReading = isReading;
            this.view = view;
        }

        public void Run()
        {
            if (isReading)
                ReadFromFile();
            else
                WriteToFile();
        }

        // Reads employee data from the file.
        // Demonstrates exception handling using try-catch blocks.
        private void ReadFromFile()
        {
            view.ShowMessage($"[THREAD-{Thread.CurrentThread.ManagedThreadId}] Reading from file: {filename}");
            try
            {
                if (File.Exists(filename))
                {
                    string[] lines = File.ReadAllLines(filename);
                    foreach (string line in lines)
                    {
                        view.ShowMessage($"Read: {line}");
                    }
                    view.ShowMessage("[THREAD] File reading completed successfully");
                }
                else
                {
                    view.ShowMessage("[THREAD] File does not exist yet");
                }
            }
            catch (IOException e)
            {
                // Handles file-related errors during writing.
                view.ShowError($"IO Error while reading: {e.Message}");
            }
        }

        private void WriteToFile()
        {
            view.ShowMessage($"[THREAD-{Thread.CurrentThread.ManagedThreadId}] Writing to file: {filename}");
            try
            {
                using (StreamWriter writer = new StreamWriter(filename))
                {
                    foreach (Employee emp in employees)
                    {
                        writer.WriteLine(emp.ToString());
                    }
                }
                view.ShowMessage($"[THREAD] File writing completed successfully. {employees.Count} records saved.");
            }
            catch (IOException e)
            {
                view.ShowError($"IO Error while writing: {e.Message}");
            }
        }
    }
}
